docclean/cycle_gan.py

```diff
@@ -0,0 +1,147 @@
+
+        if self.restor_checpoint:
+            # if a checkpoint exists, restore the latest checkpoint.
+            if self.ckpt_manager.latest_checkpoint:
+                self.ckpt.restore(self.ckpt_manager.latest_checkpoint)
+                print('Latest checkpoint restored!!')
+
+    def discriminator_loss(self, real, generated):
+        real_loss = self.loss_obj(tf.ones_like(real), real)
+        generated_loss = self.loss_obj(tf.zeros_like(generated), generated)
+        total_disc_loss = real_loss + generated_loss
+        return total_disc_loss * 0.5
+
+    def generator_loss(self, generated):
+        return self.loss_obj(tf.ones_like(generated), generated)
+
+    def calc_cycle_loss(self, real_image, cycled_image):
+        loss1 = tf.reduce_mean(tf.abs(real_image - cycled_image))
+        return LAMBDA * loss1
+
+    def identity_loss(self, real_image, same_image):
+        loss = tf.reduce_mean(tf.abs(real_image - same_image))
+        return LAMBDA * 0.5 * loss
+
+    @tf.function
+    def train_step(self, real_x, real_y):
+        # persistent is set to True because the tape is used more than
+        # once to calculate the gradients.
+        with tf.GradientTape(persistent=True) as tape:
+            # Generator G translates X -> Y
+            # Generator F translates Y -> X.
+
+            fake_y = self.generator_g(real_x, training=True)
+            cycled_x = self.generator_f(fake_y, training=True)
+
+            fake_x = self.generator_f(real_y, training=True)
+            cycled_y = self.generator_g(fake_x, training=True)
+
+            # same_x and same_y are used for identity loss.
+            same_x = self.generator_f(real_x, training=True)
+            same_y = self.generator_g(real_y, training=True)
+
+            disc_real_x = self.discriminator_x(real_x, training=True)
+            disc_real_y = self.discriminator_y(real_y, training=True)
+
+            disc_fake_x = self.discriminator_x(fake_x, training=True)
+            disc_fake_y = self.discriminator_y(fake_y, training=True)
+
+            # calculate the loss
+            gen_g_loss = self.generator_loss(disc_fake_y)
+            gen_f_loss = self.generator_loss(disc_fake_x)
+
+            total_cycle_loss = self.calc_cycle_loss(real_x, cycled_x) + self.calc_cycle_loss(real_y, cycled_y)
+
+            # Total generator loss = adversarial loss + cycle loss
+            total_gen_g_loss = gen_g_loss + total_cycle_loss + self.identity_loss(real_y, same_y)
+            total_gen_f_loss = gen_f_loss + total_cycle_loss + self.identity_loss(real_x, same_x)
+
+            disc_x_loss = self.discriminator_loss(disc_real_x, disc_fake_x)
+            disc_y_loss = self.discriminator_loss(disc_real_y, disc_fake_y)
+
+        # Calculate the gradients for generator and discriminator
+        generator_g_gradients = tape.gradient(total_gen_g_loss,
+                                              generator_g.trainable_variables)
```

Comment:
in the following lines, I believe `generator_g`, `generator_f`, `discriminator_x`, `discriminator_y` should all have `self.`

Reply:
Fixed in 92ee6ee.